pkg/controller/security/manager.go

```diff
@@ -0,0 +1,289 @@
+
+	// storing certificates
+	certsCache *certsCache
+
+	// Prioritize certificates based on exp
+	certsQueue *certificateQueue
+
+	certsOpChan chan securityData
+}
+
+type certificateQueue struct {
+	certs []*certExp
+	mu    sync.Mutex
+}
+
+func (s *SecretManager) SendData(identity string, op int) {
+	data := securityData{Identity: identity, Operation: op}
+	s.certsOpChan <- data
+}
+
+func (s *SecretManager) operateCerts() {
+	for data := range s.certsOpChan {
+		identity, op := data.Identity, data.Operation
+
+		switch op {
+		case ApplyCert:
+			go s.addCerts(identity)
+		case DeleteCert:
+			go s.deleteCerts(identity)
+		}
+	}
+}
+
+func newCertCache() *certsCache {
+	return &certsCache{
+		certs: make(map[string]*certCache),
+		mu:    sync.Mutex{},
+	}
+}
+
+func newPriorityQueue() *certificateQueue {
+	return &certificateQueue{
+		certs: make([]*certExp, 0),
+		mu:    sync.Mutex{},
+	}
+}
+
+func (pq *certificateQueue) Push(x interface{}) {
+	n := len(pq.certs)
+	item := x.(*certExp)
+	item.index = n
+	pq.certs = append(pq.certs, item)
+}
+
+func (pq *certificateQueue) Pop() interface{} {
+	old := pq.certs
+	n := len(old)
+	x := old[n-1]
+	old[n-1] = nil // avoid memory leak
+	x.index = -1
+	pq.certs = old[0 : n-1]
+	return x
+}
+
+func (pq *certificateQueue) Len() int {
+	return len(pq.certs)
+}
+
+func (pq *certificateQueue) Less(i, j int) bool {
+	return pq.certs[i].exp.Before(pq.certs[j].exp)
+}
+
+func (pq *certificateQueue) Swap(i, j int) {
+	pq.certs[i], pq.certs[j] = pq.certs[j], pq.certs[i]
+}
+
+func (pq *certificateQueue) addItem(certExp *certExp) {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	heap.Push(pq, certExp)
+}
+
+func (pq *certificateQueue) delete(identity string) *certExp {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	for i := 0; i < len(pq.certs); i++ {
+		if pq.certs[i].identity == identity {
+			return heap.Remove(pq, i).(*certExp)
+		}
+	}
+	return nil
+}
+
+// Find the top item and pop if it is about to expire
+func (pq *certificateQueue) lookupAndPop() *certExp {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	top := pq.certs[0]
+	if time.Until(top.exp.Add(-10*time.Minute)) <= 0 {
+		return heap.Pop(pq).(*certExp)
+	}
+	return nil
+}
+
+func (s *SecretManager) lookupAndStoreCert(identity string, newCert *istiosecurity.SecretItem) {
+	s.certsCache.mu.Lock()
+	defer s.certsCache.mu.Unlock()
+	// Check if the key exists in the map
+	// If refCnt == 0, then this certificate is about to be deleted, so do not perform a refresh.
+	certCache := s.certsCache.certs[identity]
+	if certCache != nil && certCache.refCnt != 0 {
+		certCache.cert = newCert
+		certExp := certExp{exp: newCert.ExpireTime, identity: identity}
+		s.certsQueue.addItem(&certExp)
+		log.Infof("cert %v add, exp:%v\n", identity, newCert.ExpireTime)
+	}
+}
+
+// Check if the certificate already exists, if it does, increment the reference count by 1,
+// if it doesn't, request a new certificate.
+func (c *certsCache) lookupOrStore(identity string) *certCache {
```

Comment:
```suggestion
func (c *certsCache) addOrUpdate(identity string) *certCache {
```

Reply:
I think we donot need a return value , so remove *certCache

Reply:
ok